operator/pkg/lbipam/lbipam.go

```diff
@@ -0,0 +1,1797 @@
+	}
+
+	err = ipam.satisfyAndUpdateCounts()
+	if err != nil {
+		ipam.logger.WithError(err).Error("Error in satisfyAndUpdateCounts()")
+		return fmt.Errorf("satisfyAndUpdateCounts: %w", err)
+	}
+
+	return nil
+}
+
+func (ipam *LBIPAM) poolOnDelete(k resource.Key, pool *cilium_api_v2alpha1.CiliumLoadBalancerIPPool) error {
+	err := ipam.handlePoolDeleted(pool)
+	if err != nil {
+		ipam.logger.WithError(err).Error("Error in handlePoolDeleted()")
+		return fmt.Errorf("handlePoolDeleted: %w", err)
+	}
+
+	err = ipam.settleConflicts()
+	if err != nil {
+		ipam.logger.WithError(err).Error("Error in settleConflicts()")
+		return fmt.Errorf("settleConflicts: %w", err)
+	}
+
+	err = ipam.satisfyAndUpdateCounts()
+	if err != nil {
+		ipam.logger.WithError(err).Error("Error in satisfyAndUpdateCounts()")
+		return fmt.Errorf("satisfyAndUpdateCounts: %w", err)
+	}
+
+	return nil
+}
+
+func (ipam *LBIPAM) svcOnUpsert(k resource.Key, svc *slim_core_v1.Service) error {
+	// Deep copy so we get a version we are allowed to update
+	svc = svc.DeepCopy()
+
+	ipam.logger.Debugf("Added/updated service '%s/%s'", svc.GetNamespace(), svc.GetName())
+
+	err := ipam.handleUpsertService(svc)
+	if err != nil {
+		ipam.logger.WithError(err).Error("Error in handleUpsertService()")
+		return fmt.Errorf("handleUpsertService: %w", err)
+	}
+
+	err = ipam.satisfyAndUpdateCounts()
+	if err != nil {
+		ipam.logger.WithError(err).Error("Error in satisfyAndUpdateCounts()")
+		return fmt.Errorf("satisfyAndUpdateCounts: %w", err)
+	}
+
+	return nil
+}
+
+func (ipam *LBIPAM) svcOnDelete(k resource.Key, svc *slim_core_v1.Service) error {
+	// Deep copy so we get a version we are allowed to update
+	svc = svc.DeepCopy()
+
+	ipam.logger.Debugf("Deleted service '%s/%s'", svc.GetNamespace(), svc.GetName())
+
+	err := ipam.handleDeletedService(svc)
+	if err != nil {
+		ipam.logger.WithError(err).Error("Error in handleDeletedService()")
+		return fmt.Errorf("handleDeletedService: %w", err)
+	}
+
+	err = ipam.satisfyAndUpdateCounts()
+	if err != nil {
+		ipam.logger.WithError(err).Error("Error in satisfyAndUpdateCounts()")
+		return fmt.Errorf("satisfyAndUpdateCounts: %w", err)
+	}
+
+	return nil
+}
+
+func (ipam *LBIPAM) satisfyAndUpdateCounts() error {
+	err := ipam.satisfyServices()
+	if err != nil {
+		return fmt.Errorf("satisfyServices: %w", err)
+	}
+
+	err = ipam.updateAllPoolCounts()
+	if err != nil {
+		return fmt.Errorf("updateAllPoolCounts: %w", err)
+	}
+
+	return nil
+}
+
+// handleUpsertService updates the service view in the service store, it removes any allocation and ingress that
+// do not belong on the service and will move the service to the satisfied or unsatisfied service view store depending
+// on if the service requests are satisfied or not.
+func (ipam *LBIPAM) handleUpsertService(svc *slim_core_v1.Service) error {
+	key := resource.NewKey(svc)
+	sv, found, _ := ipam.serviceStore.GetService(key)
+	if !found {
+		sv = &ServiceView{
+			Key: key,
+		}
+	}
```

Comment:
Here would be the place where we deepcopy the `sv` if we found it in the `ipam.serviceStore`

Reply:
`sv` is a ServiceView object, which is owned by the LB-IPAM controller, no external references exist, so there is no need to perform a deep copy here.

Reply:
I thought this was a k8s store.